hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/TrashOzoneFileSystem.java

```diff
@@ -0,0 +1,422 @@
+          }
+        }
+        return true;
+      } else {
+        LOG.trace("iterating file: {}", path);
+        keyPathList.add(pathKey);
+        return processKeyPath(keyPathList);
+      }
+    }
+
+    FileStatus getStatus() {
+      return status;
+    }
+  }
+
+  private class RenameIterator extends OzoneListingIterator {
+    private final String srcPath;
+    private final String dstPath;
+
+    RenameIterator(Path srcPath, Path dstPath)
+        throws IOException {
+      super(srcPath);
+      this.srcPath = pathToKey(srcPath);
+      this.dstPath = pathToKey(dstPath);
+      LOG.trace("rename from:{} to:{}", this.srcPath, this.dstPath);
+    }
+
+    @Override
+    boolean processKeyPath(List<String> keyPathList) {
+      for (String keyPath : keyPathList) {
+        String newPath = dstPath.concat(keyPath.substring(srcPath.length()));
+        OFSPath src = new OFSPath(keyPath);
+        OFSPath dst = new OFSPath(newPath);
+
+        String volumeName = src.getVolumeName();
+        String bucketName = src.getBucketName();
+        String keyName = src.getKeyName();
+
+        OzoneManagerProtocolProtos.KeyArgs keyArgs =
+            OzoneManagerProtocolProtos.KeyArgs.newBuilder().setKeyName(keyName)
+            .setVolumeName(volumeName).setBucketName(bucketName).build();
+
+        String toKeyName = dst.getKeyName();
+
+        OzoneManagerProtocolProtos.RenameKeyRequest renameKeyRequest =
+            OzoneManagerProtocolProtos.RenameKeyRequest.newBuilder()
+            .setKeyArgs(keyArgs).setToKeyName(toKeyName).build();
+        OzoneManagerProtocolProtos.OMRequest omRequest =
+            OzoneManagerProtocolProtos.OMRequest.newBuilder()
+            .setClientId(UUID.randomUUID().toString())
+            .setRenameKeyRequest(renameKeyRequest)
+            .setCmdType(OzoneManagerProtocolProtos.Type.RenameKey).build();
+        try {
+          ozoneManager.getOmServerProtocol().
+              submitRequest(NULL_RPC_CONTROLLER, omRequest);
+        } catch (ServiceException e) {
+          LOG.error("Couldn't send rename request.");
+        }
+
+      }
+      return true;
+    }
+  }
+
+  private class DeleteIterator extends OzoneListingIterator {
+    final private boolean recursive;
+
+
+    DeleteIterator(Path f, boolean recursive)
+        throws IOException {
+      super(f);
+      this.recursive = recursive;
+      if (getStatus().isDirectory()
+          && !this.recursive
+          && listStatus(f).length != 0) {
+        throw new PathIsNotEmptyDirectoryException(f.toString());
+      }
+    }
+
+    @Override
+    boolean processKeyPath(List<String> keyPathList) {
+      LOG.trace("Deleting keys: {}", keyPathList);
+
+      String volumeName;
+      String bucketName;
+      List<String> keyList = keyPathList.stream()
+          .map(p -> new OFSPath(p).getKeyName())
+          .collect(Collectors.toList());
+
+      if(!keyPathList.isEmpty()){
+        OFSPath p = new OFSPath(keyPathList.get(0));
```

Comment:
why are we deleting only the first entry ? and ignoring the rest of the entries here ?

Reply:
the keys in the keyset belong to the same volume and bucket.hence just reading the volume and bucket from the first entry. we are not ignoring other entries. all entries are deleted .

Reply:
Sure. got it thanks.